### PR TITLE
Pass the whole VPA into cappingRecommendationProcessor.Apply()

### LINKED ISSUE #7491: VPA wrongly determine container name?
vpa-autoscaler-updater pod can't adjust cert-manager deployment, there is an error:

1113 12:33:40.225140       1 capping.go:79] no matching Container found for recommendation cert-manager

vpa helm chart: 9.9.0

vpa.cert-manager:
```
Name:         cert-manager
Namespace:    cert-manager
Labels:       kustomize.toolkit.fluxcd.io/name=cert-manager-release
              kustomize.toolkit.fluxcd.io/namespace=flux-system
Annotations:  <none>
API Version:  autoscaling.k8s.io/v1
Kind:         VerticalPodAutoscaler
Metadata:
  Creation Timestamp:  2023-10-18T14:14:12Z
  Generation:          56939
  Resource Version:    1851559923
  UID:                 adf8e2f7-13b4-438b-9fd8-d560212f9bfd
Spec:
  Target Ref:
    API Version:  apps/v1
    Kind:         Deployment
    Name:         cert-manager
  Update Policy:
    Min Replicas:  1
    Update Mode:   Recreate
Status:
  Conditions:
    Last Transition Time:  2023-10-18T14:14:41Z
    Status:                True
    Type:                  RecommendationProvided
  Recommendation:
    Container Recommendations:
      Container Name:  cert-manager-controller
      Lower Bound:
        Cpu:     25m
        Memory:  49566182
      Target:
        Cpu:     25m
        Memory:  49566436
      Uncapped Target:
        Cpu:     25m
        Memory:  49566436
      Upper Bound:
        Cpu:     25m
        Memory:  49693235
Events:          <none>
``` 

cert-manager deployment:
```
Name:                   cert-manager
Namespace:              cert-manager
CreationTimestamp:      Wed, 18 Oct 2023 17:14:32 +0300
Labels:                 app=cert-manager
                        app.kubernetes.io/component=controller
                        app.kubernetes.io/instance=cert-manager
                        app.kubernetes.io/managed-by=Helm
                        app.kubernetes.io/name=cert-manager
                        app.kubernetes.io/version=v1.16.1
                        helm.sh/chart=cert-manager-v1.16.1
                        helm.toolkit.fluxcd.io/name=cert-manager
                        helm.toolkit.fluxcd.io/namespace=cert-manager
Annotations:            deployment.kubernetes.io/revision: 9
                        meta.helm.sh/release-name: cert-manager
                        meta.helm.sh/release-namespace: cert-manager
Selector:               app.kubernetes.io/component=controller,app.kubernetes.io/instance=cert-manager,app.kubernetes.io/name=cert-manager
Replicas:               1 desired | 1 updated | 1 total | 1 available | 0 unavailable
StrategyType:           RollingUpdate
MinReadySeconds:        0
RollingUpdateStrategy:  25% max unavailable, 25% max surge
Pod Template:
  Labels:           app=cert-manager
                    app.kubernetes.io/component=controller
                    app.kubernetes.io/instance=cert-manager
                    app.kubernetes.io/managed-by=Helm
                    app.kubernetes.io/name=cert-manager
                    app.kubernetes.io/version=v1.16.1
                    helm.sh/chart=cert-manager-v1.16.1
  Service Account:  certmanager-oidc
  Containers:
   cert-manager-controller:
    Image:       quay.io/jetstack/cert-manager-controller:v1.16.1
    Ports:       9402/TCP, 9403/TCP
    Host Ports:  0/TCP, 0/TCP
    Args:
      --v=2
      --cluster-resource-namespace=$(POD_NAMESPACE)
      --leader-election-namespace=kube-system
      --acme-http01-solver-image=quay.io/jetstack/cert-manager-acmesolver:v1.16.1
      --default-issuer-name=letsencrypt-prod
      --default-issuer-kind=ClusterIssuer
      --max-concurrent-challenges=60
    Limits:
      memory:  384Mi
    Requests:
      cpu:     10m
      memory:  160Mi
    Liveness:  http-get http://:http-healthz/livez delay=10s timeout=15s period=10s #success=1 #failure=8
    Environment:
      POD_NAMESPACE:     (v1:metadata.namespace)
    Mounts:             <none>
  Volumes:              <none>
  Priority Class Name:  above-average
  Node-Selectors:       kubernetes.io/os=linux
  Tolerations:          arch=arm64:NoSchedule
Conditions:
  Type           Status  Reason
  ----           ------  ------
  Available      True    MinimumReplicasAvailable
  Progressing    True    NewReplicaSetAvailable
OldReplicaSets:  cert-manager-7d47d666f8 (0/0 replicas created), cert-manager-6c59597cc8 (0/0 replicas created), cert-manager-6dc64d89dd (0/0 replicas created), cert-manager-69d7786674 (0/0 replicas created), cert-manager-778b689cb4 (0/0 replicas created), cert-manager-fd64d8c78 (0/0 replicas created), cert-manager-656d494f58 (0/0 replicas created), cert-manager-77fd697866 (0/0 replicas created)
NewReplicaSet:   cert-manager-d69584748 (1/1 replicas created)
Events:          <none>
``` 

vpa HR:
```
Name:         vertical-pod-autoscaler
Namespace:    kube-system
Labels:       kustomize.toolkit.fluxcd.io/name=vertical-pod-autoscaler
              kustomize.toolkit.fluxcd.io/namespace=flux-system
Annotations:  reconcile.fluxcd.io/requestedAt: 2024-07-31T07:25:08.69033819Z
API Version:  helm.toolkit.fluxcd.io/v2
Kind:         HelmRelease
Metadata:
  Creation Timestamp:  2023-10-18T14:10:36Z
  Finalizers:
    finalizers.fluxcd.io
  Generation:        10
  Resource Version:  1851559957
  UID:               e3126d59-4cb0-4e74-a0a1-91d728a1e18a
Spec:
  Chart:
    Spec:
      Chart:               vertical-pod-autoscaler
      Interval:            14m
      Reconcile Strategy:  ChartVersion
      Source Ref:
        Kind:       HelmRepository
        Name:       vertical-pod-autoscaler
        Namespace:  kube-system
      Version:      9.9.0
  Install:
    Crds:              CreateReplace
    Create Namespace:  true
  Interval:            9m
  Release Name:        vertical-pod-autoscaler
  Upgrade:
    Crds:  CreateReplace
  Values:
    Admission Controller:
      Affinity:
        Node Affinity:
          Preferred During Scheduling Ignored During Execution:
            Preference:
              Match Expressions:
                Key:       kubernetes.io/arch
                Operator:  In
                Values:
                  arm64
            Weight:  1
          Required During Scheduling Ignored During Execution:
            Node Selector Terms:
              Match Expressions:
                Key:       kubernetes.io/os
                Operator:  In
                Values:
                  linux
      Resources:
        Limits:
          Memory:  500Mi
        Requests:
          Cpu:     10m
          Memory:  40Mi
      Security Context:
        Allow Privilege Escalation:  false
        Capabilities:
          Drop:
            ALL
        Read Only Root Filesystem:  true
        Run As Non Root:            true
        Run As User:                65534
        Seccomp Profile:
          Type:  RuntimeDefault
      Tolerations:
        Effect:    NoSchedule
        Key:       arch
        Operator:  Equal
        Value:     arm64
    Recommender:
      Affinity:
        Node Affinity:
          Preferred During Scheduling Ignored During Execution:
            Preference:
              Match Expressions:
                Key:       kubernetes.io/arch
                Operator:  In
                Values:
                  arm64
            Weight:  1
          Required During Scheduling Ignored During Execution:
            Node Selector Terms:
              Match Expressions:
                Key:       kubernetes.io/os
                Operator:  In
                Values:
                  linux
      Extra Args:
        Pod - Recommendation - Min - Memory - Mb:  30
      Resources:
        Limits:
          Memory:  250Mi
        Requests:
          Cpu:     10m
          Memory:  150Mi
      Security Context:
        Allow Privilege Escalation:  false
        Capabilities:
          Drop:
            ALL
        Read Only Root Filesystem:  true
        Run As Non Root:            true
        Run As User:                65534
        Seccomp Profile:
          Type:  RuntimeDefault
      Tolerations:
        Effect:    NoSchedule
        Key:       arch
        Operator:  Equal
        Value:     arm64
    Updater:
      Affinity:
        Node Affinity:
          Preferred During Scheduling Ignored During Execution:
            Preference:
              Match Expressions:
                Key:       kubernetes.io/arch
                Operator:  In
                Values:
                  arm64
            Weight:  1
          Required During Scheduling Ignored During Execution:
            Node Selector Terms:
              Match Expressions:
                Key:       kubernetes.io/os
                Operator:  In
                Values:
                  linux
      Resources:
        Limits:
          Memory:  500Mi
        Requests:
          Cpu:     10m
          Memory:  150Mi
      Security Context:
        Allow Privilege Escalation:  false
        Capabilities:
          Drop:
            ALL
        Read Only Root Filesystem:  true
        Run As Non Root:            true
        Run As User:                65534
        Seccomp Profile:
          Type:  RuntimeDefault
      Tolerations:
        Effect:    NoSchedule
        Key:       arch
        Operator:  Equal
        Value:     arm64
Status:
  Conditions:
    Last Transition Time:  2024-11-13T07:13:20Z
    Message:               Helm upgrade succeeded for release kube-system/vertical-pod-autoscaler.v11 with chart vertical-pod-autoscaler@9.9.0
    Observed Generation:   10
    Reason:                UpgradeSucceeded
    Status:                True
    Type:                  Ready
    Last Transition Time:  2024-11-13T07:13:20Z
    Message:               Helm upgrade succeeded for release kube-system/vertical-pod-autoscaler.v11 with chart vertical-pod-autoscaler@9.9.0
    Observed Generation:   10
    Reason:                UpgradeSucceeded
    Status:                True
    Type:                  Released
  Helm Chart:              kube-system/kube-system-vertical-pod-autoscaler
  History:
    App Version:                  1.2.1
    Chart Name:                   vertical-pod-autoscaler
    Chart Version:                9.9.0
    Config Digest:                sha256:f6966be8af4d78d11c8766b812e13a5a7b5780ab71b83ca56c9b5c1c4117d542
    Digest:                       sha256:5cfeac276ecd02c5017829d95212348723a8d13b1f23a1968e0e0fa19e6ca2c8
    First Deployed:               2023-10-18T14:10:38Z
    Last Deployed:                2024-11-13T07:12:32Z
    Name:                         vertical-pod-autoscaler
    Namespace:                    kube-system
    Status:                       deployed
    Version:                      11
    App Version:                  1.1.2
    Chart Name:                   vertical-pod-autoscaler
    Chart Version:                9.8.3
    Config Digest:                sha256:f6966be8af4d78d11c8766b812e13a5a7b5780ab71b83ca56c9b5c1c4117d542
    Digest:                       sha256:194f51454240a7b164223ba3092ee254f61a446d5b44dd26f708b7d992fcf331
    First Deployed:               2023-10-18T14:10:38Z
    Last Deployed:                2024-07-31T07:16:25Z
    Name:                         vertical-pod-autoscaler
    Namespace:                    kube-system
    Status:                       superseded
    Version:                      10
  Last Attempted Config Digest:   sha256:f6966be8af4d78d11c8766b812e13a5a7b5780ab71b83ca56c9b5c1c4117d542
  Last Attempted Generation:      10
  Last Attempted Release Action:  upgrade
  Last Attempted Revision:        9.9.0
  Last Handled Reconcile At:      2024-07-31T07:25:08.69033819Z
  Observed Generation:            10
  Storage Namespace:              kube-system
Events:                           <none>

``` 

related to [https://github.com/kubernetes/autoscaler/issues/6215](url)

### PR DESCRIPTION
**Pull request by adrianmoisey**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

The current log message for when no container is found is very misleading and can cause confusion.

This passes the entire VPA object into that function, in order for it to create a log file with the relevant VPA name in it.

It kinda feels like surgery with a scalpel, any alternative approaches would be appreciated.

Take a look at https://github.com/kubernetes/autoscaler/issues/7491 to see the confusion it can cause.

#### Which issue(s) this PR fixes:

Fixes #7491

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: vertical-pod-autoscaler/pkg/utils/vpa/capping.go [hunk 1]
```diff
@@ -52,12 +52,13 @@ type cappingRecommendationProcessor struct {
 
 // Apply returns a recommendation for the given pod, adjusted to obey policy and limits.
 func (c *cappingRecommendationProcessor) Apply(
-	podRecommendation *vpa_types.RecommendedPodResources,
-	policy *vpa_types.PodResourcePolicy,
-	conditions []vpa_types.VerticalPodAutoscalerCondition,
+	vpa *vpa_types.VerticalPodAutoscaler,
 	pod *apiv1.Pod) (*vpa_types.RecommendedPodResources, ContainerToAnnotationsMap, error) {
 	// TODO: Annotate if request enforced by maintaining proportion with limit and allowed limit range is in conflict with policy.
```
Comment (omerap12) [inline vertical-pod-autoscaler/pkg/utils/vpa/capping.go:58]: Can we check for edge cases ( just to avoid panic )?
```go
if vpa == nil {
    return nil, nil, fmt.Errorf("cannot process nil vpa")
    }
if pod == nil {
    return nil, nil, fmt.Errorf("cannot process nil pod")
    }
```

File: vertical-pod-autoscaler/pkg/utils/vpa/capping.go [hunk 2]
```diff
@@ -52,12 +52,13 @@ type cappingRecommendationProcessor struct {
...
 
+	policy := vpa.Spec.ResourcePolicy.DeepCopy()
```
Comment (omerap12) [inline vertical-pod-autoscaler/pkg/utils/vpa/capping.go:59]: Can we check for edge cases ( just to avoid panic )?
```go
if vpa == nil {
    return nil, nil, fmt.Errorf("cannot process nil vpa")
    }
if pod == nil {
    return nil, nil, fmt.Errorf("cannot process nil pod")
    }
if vpa.Status.Recommendation == nil {
        return nil, nil, nil  // This matches existing behavior for no recommendation
    }
```

File: vertical-pod-autoscaler/pkg/utils/vpa/sequential_processor.go [hunk 3]
```diff
@@ -31,19 +31,19 @@ type sequentialRecommendationProcessor struct {
 }
 
 // Apply chains calls to underlying RecommendationProcessors in order provided on object construction
-func (p *sequentialRecommendationProcessor) Apply(podRecommendation *vpa_types.RecommendedPodResources,
-	policy *vpa_types.PodResourcePolicy,
-	conditions []vpa_types.VerticalPodAutoscalerCondition,
+func (p *sequentialRecommendationProcessor) Apply(vpa *vpa_types.VerticalPodAutoscaler,
 	pod *v1.Pod) (*vpa_types.RecommendedPodResources, ContainerToAnnotationsMap, error) {
-	recommendation := podRecommendation
+
```
Comment (omerap12) [inline vertical-pod-autoscaler/pkg/utils/vpa/sequential_processor.go:36]: Same for here ( not sure if needed but just my opinion ):
```go
if vpa == nil {
      return nil, nil, fmt.Errorf("cannot process nil vpa")
  }
if vpa.Status.Recommendation == nil {
    return nil, nil, nil
}
```

**Review by omerap12 — COMMENTED** [review]
If my comments make sense to you, we could add more unit tests to cover those edge cases. For example:
```go
func TestApplyWithNilVPA(t *testing.T) {
    pod := test.Pod().WithName("pod1").AddContainer(test.Container().WithName("ctr-name").Get()).Get()
    processor := NewCappingRecommendationProcessor(&fakeLimitRangeCalculator{})
    
    res, annotations, err := processor.Apply(nil, pod)
    assert.Error(t, err)
    assert.Nil(t, res)
    assert.Nil(t, annotations)
}

func TestApplyWithNilPod(t *testing.T) {
    vpa := test.VerticalPodAutoscaler().WithContainer("container").Get()
    processor := NewCappingRecommendationProcessor(&fakeLimitRangeCalculator{})
    
    res, annotations, err := processor.Apply(vpa, nil)
    assert.Error(t, err)
    assert.Nil(t, res)
    assert.Nil(t, annotations)
}

**Comment by raywainman** [comment]
Nice cleanup, thank you :)

/lgtm

**Review by raywainman — COMMENTED** [review]
Sorry realized I didn't submit my comments

**Comment by adrianmoisey** [comment]
Thanks for the reviews. I was actually rushed when making this, so still need to do a cleanup and respond to comments.
I'll hold for now until I can address all of that
/hold

File: vertical-pod-autoscaler/pkg/utils/vpa/capping.go (hunk 2 above)
Comment (adrianmoisey) [reply vertical-pod-autoscaler/pkg/utils/vpa/capping.go]: Should be fixed in 13d6ffa28

File: vertical-pod-autoscaler/pkg/utils/vpa/sequential_processor.go (hunk 3 above)
Comment (adrianmoisey) [reply vertical-pod-autoscaler/pkg/utils/vpa/sequential_processor.go:39]: Should be fixed in 13d6ffa28

Comment (raywainman) [reply vertical-pod-autoscaler/pkg/utils/vpa/sequential_processor.go:39]: Should we check pod here as well?

File: vertical-pod-autoscaler/pkg/utils/vpa/sequential_processor.go [hunk 4]
```diff
@@ -31,19 +33,28 @@ type sequentialRecommendationProcessor struct {
...
+func (p *sequentialRecommendationProcessor) Apply(
+	vpa *vpa_types.VerticalPodAutoscaler,
 	pod *v1.Pod) (*vpa_types.RecommendedPodResources, ContainerToAnnotationsMap, error) {
-	recommendation := podRecommendation
+
+	if vpa == nil {
+		return nil, nil, fmt.Errorf("cannot process nil vpa")
+	}
+	if vpa.Status.Recommendation == nil {
+		return nil, nil, nil
```
Comment (raywainman) [inline vertical-pod-autoscaler/pkg/utils/vpa/sequential_processor.go:44]: I actually wonder if this should be an error case...

The processors are generally "adjusting" the recommendation one at a time. If there is no recommendation to adjust then this feels like an error case to me.

@omerap12 or @adrianmoisey are you seeing a case where this should be allowed?

Comment (raywainman) [reply vertical-pod-autoscaler/pkg/utils/vpa/sequential_processor.go:44]: Oh wait I see that this is being done in the capping processor. Perhaps it is best to be consistent.

However I wonder if this can eventually be made more strict and turned into an error.

**Comment by adrianmoisey** [comment]
> Should we check pod here as well?

For some reason I can't reply in the thread where you made this comment.

I thought about adding a check there, but all that function does is pass Pod on to another function call. 
Since this function doesn't actually do anything to Pod, I figured the guards could be in the functions that would do something with Pod.

File: vertical-pod-autoscaler/pkg/utils/vpa/sequential_processor.go (hunk 4 above)
Comment (adrianmoisey) [reply vertical-pod-autoscaler/pkg/utils/vpa/sequential_processor.go:44]: I am assuming that this case should never happen

**Review by voelzmo — CHANGES_REQUESTED** [review]
Thanks, this change really improves the error message and should be a good step forward to improve the situation. I have just some small comments where I don't want to block the approval on.

Looking at this I'm reminded that we have https://github.com/kubernetes/autoscaler/issues/6744 still open and never got to reviewing https://github.com/kubernetes/autoscaler/pull/6745 to the end. This should fix the main reason for seeing these messages.

File: vertical-pod-autoscaler/pkg/utils/vpa/recommendation_processor.go [hunk 5]
```diff
@@ -29,8 +29,6 @@ type RecommendationProcessor interface {
 	// Apply processes and updates recommendation for given pod, based on container limits,
 	// VPA policy and possibly other internal RecommendationProcessor context.
 	// Must return a non-nil pointer to RecommendedPodResources or error.
-	Apply(podRecommendation *vpa_types.RecommendedPodResources,
-		policy *vpa_types.PodResourcePolicy,
-		conditions []vpa_types.VerticalPodAutoscalerCondition,
+	Apply(Vpa *vpa_types.VerticalPodAutoscaler,
```
[the remark this answers is not in the document]
Comment (adrianmoisey) [reply vertical-pod-autoscaler/pkg/utils/vpa/recommendation_processor.go]: Fixed in 3ea36d282b6c64cb6f7142f25d3855d8ef995e94

File: vertical-pod-autoscaler/pkg/utils/vpa/capping.go [hunk 6]
```diff
@@ -52,12 +52,20 @@ type cappingRecommendationProcessor struct {
...
 // Apply returns a recommendation for the given pod, adjusted to obey policy and limits.
 func (c *cappingRecommendationProcessor) Apply(
-	podRecommendation *vpa_types.RecommendedPodResources,
-	policy *vpa_types.PodResourcePolicy,
-	conditions []vpa_types.VerticalPodAutoscalerCondition,
+	vpa *vpa_types.VerticalPodAutoscaler,
 	pod *apiv1.Pod) (*vpa_types.RecommendedPodResources, ContainerToAnnotationsMap, error) {
 	// TODO: Annotate if request enforced by maintaining proportion with limit and allowed limit range is in conflict with policy.
 
+	if vpa == nil {
```
[the remark this answers is not in the document]
Comment (adrianmoisey) [reply vertical-pod-autoscaler/pkg/utils/vpa/capping.go:59]: That is something I had considered, but I landed on the fact that this function should be able to handle any possible input.

I don't actually know if there's a Go or Kubernetes best practice around this. I figured that more defensive code is a good thing, but I'm also happy to change, but I'll wait for consensus before changing it.

[the remark this answers is not in the document]
Comment (omerap12) [reply vertical-pod-autoscaler/pkg/utils/vpa/capping.go:59]: In my opinion, since this is an exported function, we should include the nil check as to prevent panics from invalid usage. But, I do agree that we're only changing the signature to so it's unlikely the VPA object would ever be nil in current usage patterns. I'm comfortable with whatever you decide.

[the remark this answers is not in the document]
Comment (raywainman) [reply vertical-pod-autoscaler/pkg/utils/vpa/capping.go:59]: +1 for defensive approach here.

File: vertical-pod-autoscaler/pkg/utils/vpa/sequential_processor.go [hunk 7]
```diff
@@ -31,19 +33,28 @@ type sequentialRecommendationProcessor struct {
...
+	}
+
+	var recommendation *vpa_types.RecommendedPodResources
```
Comment (raywainman) [inline vertical-pod-autoscaler/pkg/utils/vpa/sequential_processor.go:47]: One small edge case here is that if there are no processors, this will be nil (where the behavior beforehand was that we would return the existing VPA recommendation).

I know we don't currently have an empty processor but should we match the behavior here?

**Comment by raywainman** [comment]
One last small comment then this is good to go, sorry for the wait.

File: vertical-pod-autoscaler/pkg/utils/vpa/sequential_processor.go (hunk 7 above)
Comment (adrianmoisey) [reply vertical-pod-autoscaler/pkg/utils/vpa/sequential_processor.go]: 🤦 Yup, good callout.
Should be fixed in 59236c93a2
